File: scripts/resource-enrichment/fmhy-enrich.js
/**
 * FMHY 模板化描述替换脚本
 * 策略：爬取资源官网的 meta description 或第一段可见文本
 * 仅替换模板化描述（包含 "An AI-powered tool" 等模板）
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');
const { URL } = require('url');

const DATA_FILE = path.join(__dirname, '../../public/data/fmhy-resources.json');
const CHECKPOINT = path.join(__dirname, 'fmhy-enrich-checkpoint.json');

// 模板化描述的特征
const TEMPLATE_PATTERNS = [
  'An AI-powered tool that helps with',
  'AI-powered',
  'helps with',
  'Useful for automating',
  'A tool that helps',
];

function isTemplateDesc(desc) {
  if (!desc) return true;
  return TEMPLATE_PATTERNS.some(p => desc.includes(p));
}

function normalizeUrl(url) {
  try {
    const u = new URL(url);
    return u.hostname.replace(/^www\./, '') + (u.pathname === '/' ? '' : u.pathname);
  } catch {
    return url;
  }
}

function fetchMetaDesc(url) {
  return new Promise((resolve) => {
    const timeout = 8000;
    const isHttps = url.startsWith('https:');
    const client = isHttps ? https : http;

    const req = client.get(url, {
      timeout,
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; CraftisleBot/1.0; +https://craftisle.com)',
        'Accept': 'text/html,application/xhtml+xml',
      },
      rejectUnauthorized: false,
    }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        const next = res.headers.location.startsWith('http')
          ? res.headers.location
          : new URL(res.headers.location, url).href;
        return resolve(fetchMetaDesc(next));
      }

      if (res.statusCode !== 200) {
        return resolve(null);
      }

      let data = '';
      let total = 0;
      const MAX = 200 * 1024; // 200KB

      res.on('data', (chunk) => {
        data += chunk.toString();
        total += chunk.length;
        if (total > MAX) {
          req.destroy();
        }
      });

      res.on('end', () => {
        resolve(extractDesc(data));
      });
    });

    req.on('timeout', () => {
      req.destroy();
      resolve(null);
    });

    req.on('error', () => resolve(null));
  });
}

function extractDesc(html) {
  // 尝试 meta description
  const metaPatterns = [
    /<meta\s+name=["']description["']\s+content=["']([^"']{20,300})["']/i,
    /<meta\s+content=["']([^"']{20,300})["']\s+name=["']description["']/i,
    /<meta\s+property=["']og:description["']\s+content=["']([^"']{20,300})["']/i,
  ];

  for (const re of metaPatterns) {
    const m = html.match(re);
    if (m && m[1].trim().length >= 20) {
      return cleanText(m[1]);
    }
  }

  // 回退：第一段可见文本
  const cleaned = html
    .replace(/<script[\s\S]*?<\/script>/gi, ' ')
    .replace(/<style[\s\S]*?<\/style>/gi, ' ')
    .replace(/<[^>]+>/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();

  if (cleaned.length > 30) {
    return cleaned.slice(0, 200).trim();
  }

  return null;
}

function cleanText(s) {
  return s.replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#\d+;/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

async function main() {
  const data = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
  const allResources = [];
  Object.values(data.categories).forEach(cat => {
    (cat.resources || []).forEach(r => allResources.push(r));
  });

  console.log(`[main] Total FMHY resources: ${allResources.length}`);

  // 找出模板化描述
  const needUpdate = allResources.filter(r => isTemplateDesc(r.description));
  console.log(`[main] Template/empty descriptions to replace: ${needUpdate.length}`);

  if (needUpdate.length === 0) {
    console.log('[main] No template descriptions found, all done!');
    return;
  }

  // 加载 checkpoint
  let checkpoint = { updated: 0, failed: 0, done: [] };
  if (fs.existsSync(CHECKPOINT)) {
    checkpoint = JSON.parse(fs.readFileSync(CHECKPOINT, 'utf8'));
    console.log(`[main] Resuming from checkpoint: updated=${checkpoint.updated}, failed=${checkpoint.failed}`);
  }

  const toProcess = needUpdate.filter(r => !checkpoint.done.includes(r.id));
  console.log(`[main] Remaining to process: ${toProcess.length}`);

  const BATCH = 30;
  const CONCURRENCY = 5;
  let idx = 0;

  async function processBatch() {
    const batch = toProcess.slice(idx, idx + BATCH);
    if (batch.length === 0) return;

    console.log(`\n[batch] Processing ${idx + 1}–${idx + batch.length} of ${toProcess.length}...`);
    idx += batch.length;

    const results = [];
    for (let i = 0; i < batch.length; i += CONCURRENCY) {
      const chunk = batch.slice(i, i + CONCURRENCY);
      const promises = chunk.map(r =>
        fetchMetaDesc(r.url)
          .then(desc => ({ id: r.id, url: r.url, desc, ok: !!desc }))
          .catch(() => ({ id: r.id, url: r.url, desc: null, ok: false }))
      );
      const settled = await Promise.all(promises);
      results.push(...settled);

      const progress = settled.filter(s => s.ok).length;
      console.log(`  chunk ${Math.floor(i / CONCURRENCY) + 1}: ${progress}/${settled.length} got description`);
    }

    // 应用结果
    results.forEach(({ id, desc, ok }) => {
      if (!ok || !desc) {
        checkpoint.failed++;
        return;
      }
      // 找到资源并更新
      for (const cat of Object.values(data.categories)) {
        const rr = (cat.resources || []).find(rr => rr.id === id);
        if (rr) {
          rr.description = desc;
          checkpoint.updated++;
          break;
        }
      }
      checkpoint.done.push(id);
    });

    // 保存 checkpoint + 写回 JSON
    fs.writeFileSync(CHECKPOINT, JSON.stringify(checkpoint, null, 2));
    fs.writeFileSync(DATA_FILE, JSON.stringify(data, null, 2), 'utf8');
    console.log(`[batch] ✅ Saved. Total updated: ${checkpoint.updated}, failed: ${checkpoint.failed}`);
  }

  while (idx < toProcess.length) {
    await processBatch();
    if (idx < toProcess.length) {
      console.log('  Waiting 3s before next batch...');
      await new Promise(r => setTimeout(r, 3000));
    }
  }

  console.log(`\n✅ All done! Updated: ${checkpoint.updated}, Failed: ${checkpoint.failed}`);
  console.log('Checkpoint saved to:', CHECKPOINT);
}

main().catch(e => console.error('Fatal:', e.message));

File: scripts/resource-enrichment/fmhy-enrich.py
#!/usr/bin/env python3
"""
FMHY 模板化描述替换脚本
- 爬取资源官网的 meta description
- 仅替换模板化/空描述
- 支持断点续传（checkpoint.json）
"""

import json, os, re, time, random
import requests
from bs4 import BeautifulSoup

DATA_FILE = os.path.join(os.path.dirname(__file__), '../../public/data/fmhy-resources.json')
CHECKPOINT = os.path.join(os.path.dirname(__file__), 'fmhy-enrich-checkpoint.json')
TEMPLATE_PATTERNS = ['An AI-powered tool', 'AI-powered', 'helps with', 'Useful for']

HEADERS = {
    'User-Agent': 'Mozilla/5.0 (compatible; CraftisleBot/1.0; +https://craftisle.com)',
    'Accept': 'text/html,application/xhtml+xml',
    'Accept-Language': 'en-US,en;q=0.9',
}

SESSION = requests.Session()
SESSION.headers.update(HEADERS)
SESSION.timeout = 10

def is_template(desc):
    if not desc or not desc.strip():
        return True
    return any(p in desc for p in TEMPLATE_PATTERNS)

def fetch_meta_desc(url):
    try:
        resp = SESSION.get(url, timeout=10, allow_redirects=True, verify=False)
        if resp.status_code != 200:
            return None
        html = resp.text[:300*1024]  # 最多300KB
        soup = BeautifulSoup(html, 'lxml')

        # 尝试 meta description
        for attr, val in [('name', 'description'), ('property', 'og:description'), ('name', 'twitter:description')]:
            tag = soup.find('meta', attrs={attr: val})
            if tag and tag.get('content', '').strip():
                content = tag['content'].strip()
                if len(content) >= 20:
                    return clean_text(content)

        # 回退：第一段可见文本
        for selector in ['p', 'article', 'section', 'main', 'div.content', 'div.description']:
            el = soup.select_one(selector)
            if el:
                text = el.get_text(strip=True)[:300]
                if len(text) >= 30:
                    return clean_text(text)

        return None
    except Exception as e:
        return None

def clean_text(s):
    return re.sub(r'\s+', ' ', s).strip()

def load_checkpoint():
    if os.path.exists(CHECKPOINT):
        return json.load(open(CHECKPOINT, 'r', encoding='utf-8'))
    return {'updated': 0, 'failed': 0, 'done': [], 'skipped': []}

def save_checkpoint(cp):
    json.dump(cp, open(CHECKPOINT, 'w', encoding='utf-8'), indent=2, ensure_ascii=False)

def main():
    with open(DATA_FILE, 'r', encoding='utf-8') as f:
        data = json.load(f)

    all_resources = []
    for cat in data.get('categories', {}).values():
        all_resources.extend(cat.get('resources', []))

    print(f'[main] Total FMHY resources: {len(all_resources)}')

    need = [r for r in all_resources if is_template(r.get('description', ''))]
    print(f'[main] Template/empty descriptions to replace: {len(need)}')

    if len(need) == 0:
        print('[main] No template descriptions found, all done!')
        return

    cp = load_checkpoint()
    to_process = [r for r in need if r['id'] not in cp['done'] and r['id'] not in cp['skipped']]
    print(f'[main] Remaining to process: {len(to_process)}')
    print(f'[main] Checkpoint: updated={cp["updated"]}, failed={cp["failed"]}')

    BATCH = 50
    idx = 0

    while idx < len(to_process):
        batch = to_process[idx:idx+BATCH]
        idx += len(batch)
        print(f'\n[batch] Processing {idx-len(batch)+1}-{idx} of {len(to_process)}...')

        for r in batch:
            desc = fetch_meta_desc(r['url'])
            if desc and len(desc) >= 20:
                # 找到并更新
                for cat in data.get('categories', {}).values():
                    rr = [x for x in cat.get('resources', []) if x['id'] == r['id']]
                    if rr:
                        rr[0]['description'] = desc
                        cp['updated'] += 1
                        cp['done'].append(r['id'])
                        print(f'  ✅ {r["name"][:40]}: {desc[:60]}...')
                        break
            else:
                cp['failed'] += 1
                cp['skipped'].append(r['id'])
                print(f'  ❌ {r["name"][:40]}: no description found')

        # 每批结束后保存
        save_checkpoint(cp)
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f'[batch] ✅ Saved. Total updated: {cp["updated"]}, failed: {cp["failed"]}')

        if idx < len(to_process):
            wait = random.uniform(2, 4)
            print(f'  Waiting {wait:.1f}s before next batch...')
            time.sleep(wait)

    print(f'\n✅ All done! Updated: {cp["updated"]}, Failed: {cp["failed"]}')
    print(f'Checkpoint saved to: {CHECKPOINT}')

if __name__ == '__main__':
    main()
